ansible/roles/docker/tasks/main.yml

```diff
@@ -38,29 +46,69 @@
   when: ansible_os_family == "RedHat"
 
 - name: check docker-tcp socket state
-  shell: systemctl status docker-tcp.socket | grep "Active:"
+  shell: systemctl status docker-tcp.socket | grep 'Active.*active' -o
   register: docker_tcp_socket_state
 
+- include: create_docker_device.yml
+  when: docker_device != ""
+
 - name: copy systemd units for docker tcp socket settings
   template: src=docker-tcp.j2 dest=/etc/systemd/system/docker-tcp.socket
   register: docker_tcp_socket
 
   # tcp socket service requires docker service to be started after it
-- name: start docker tcp socket service
-  shell: sudo systemctl stop docker && sudo systemctl start docker-tcp.socket && sudo systemctl start docker
-  when: (docker_tcp_socket | changed) or not docker_tcp_socket_state.stdout | match("Active.* active")
+- name: reload and stop docker
+  shell: sudo systemctl daemon-reload
+  when: "(docker_tcp_socket | changed) or (docker_tcp_socket_state.stdout != 'Active: active')"
+
+- name: stop docker
+  service:
+    name: docker
+    state: stopped
+  when: "(docker_tcp_socket | changed) or (docker_tcp_socket_state.stdout != 'Active: active')"
+
+- name: start docker-tcp service
+  service:
+    name: docker-tcp.socket
+    state: started
+  when: "(docker_tcp_socket | changed) or (docker_tcp_socket_state.stdout != 'Active: active')"
+
+- name: check docker service state
+  shell: sudo systemctl status docker | grep 'Active.*active' -o
+  ignore_errors: true
+  register: docker_service_state
+  tags:
+    - prebake-for-dev
 
 - name: remove the docker key file, if any. It shall be regenerated by docker on restart
   file: name=/etc/docker/key.json  state=absent
+  when: "(docker_service_state.stderr | match('.*docker.service changed on disk.*')) or (docker_service_state.stdout != 'Active: active')"
 
 # XXX: service module doesn't do daemon-reload yet, so need to use shell module here
 # https://github.com/ansible/ansible-modules-core/issues/191
-- name: restart docker
+- name: reload docker
   #service: name=docker state=restarted
-  shell: sudo systemctl daemon-reload && sudo systemctl restart docker
+  shell: sudo systemctl daemon-reload
+  when: "(docker_service_state.stderr | match('.*docker.service changed on disk.*')) or (docker_service_state.stdout != 'Active: active')"
   tags:
     - prebake-for-dev
 
+# XXX: this needs to happen twice after setting up the thin provisioner because
+#      of some docker bug I've not investigated.
+- name: restart docker (first time)
+  service:
+    name: docker
+    state: restarted
+  when: 
```

Comment:
> docker is forcefully stopped above if the conditions are met

hmm, may be I missed this, where is docker forcefully/unconditionally stopped? Unconditional stop/restart have side effects of killing running containers on replay of a playbook so we should avoid it as much possible.

I might not have been very clear earlier in my previous comment. My primary concern was that with the proposed check there is some mental deduction that needs to happen to be able to determine if this condition will do the right thing (like will it correctly re/start docker if the `docker_device` is not set?), by keeping the old condition in here ensures that it stays easy to deduce.

Also I did some testing with this patch to narrow down the problems, if any, with the proposed condition:
- this task fails if `thin_provisioned` is undefined but the error get's silently ignored as `ignore_errors` is set which I think is needed due to a docker bug but I guess we don't want the undefined variable error to be skipped.
- So I fixed that as follows and verified that it correctly get's skipped on calling `make provision`:
```
(thin_provisioned | default(false) | bool ) and
 thin_provisioned.rc == 0
```
- With the above fix, next I tried the scenario where `docker_device` is not set at time of doing `make start`. I noticed that this task is skipped i.e. docker is not at all started by this task, when I thought it should have started it.

- Changing the condition to following preserves the behavior of starting docker as today:
```
"((docker_service_state.stderr | match('.*docker.service changed on disk.*')) or
 (docker_service_state.stdout != 'Active: active')) or
((thin_provisioned | default(false) | bool) and
 thin_provisioned.rc == 0)"
```